### PR TITLE
Add PanOS Loopback Module

### PR DESCRIPTION
**Pull request by geraint-jones-nz**
This is a new PR for some old work, I have updated the module to support operation alongside state as per the new standard.

### CODE REVIEW

File: library/panos_loopback.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
+#!/usr/bin/env python
+
+#  Copyright 2016 Palo Alto Networks, Inc
```
Comment (shinmog) [inline library/panos_loopback.py]: Change this to 2019; the date is the date the file is added / created.

File: library/panos_loopback.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+DOCUMENTATION = '''
+---
+module: panos_loopback
+short_description: configure network loopback interfaces
+description:
+    - Configure loopback interfaces on PanOS
+
+
+author: "Luigi Mori (@jtschichold), Ivan Bojer (@ivanbojer), Geraint Jones (@nexus_moneky_nz)"
```
Comment (shinmog) [inline library/panos_loopback.py]: If Luigi and Ivan didn't contribute to this module, it's ok to remove their names from attribution :)

File: library/panos_loopback.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+version_added: "2.6"
```
Comment (shinmog) [inline library/panos_loopback.py]: This should be 2.8

File: library/panos_loopback.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+            - Name of the vsys (if firewall) or device group (if panorama) to put this object.
+        default: "vsys1"
+    commit:
+        description:
+            - Commit if changed
+        default: true
+'''
+
+EXAMPLES = '''
+# Create loopback.1
```
Comment (shinmog) [inline library/panos_loopback.py]: This is a deletion example.

File: library/panos_loopback.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        e = get_exception()
+        module.fail_json(msg=e.message)
+
+    # Build the object based on the user spec.
+    loopback = network.LoopbackInterface(**spec)
+    con.add(loopback)
+
+    if state == 'present':
+        if loopback.name in [x.name for x in interfaces]:
+            i = EthernetInterface(loopback.name)
```
Comment (shinmog) [inline library/panos_loopback.py]: `network.LoopbackInterface`

File: library/panos_loopback.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+            con.add(i)
+            try:
+                i.refresh()
+            except PanDeviceError as e:
+                module.fail_json(msg='Failed "present" refresh: {0}'.format(e))
+            if not i.equal(eth, compare_children=False):
```
Comment (shinmog) [inline library/panos_loopback.py]: Should be `loopback` not `eth`.

File: library/panos_loopback.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        description:
+            - Username credentials to use for auth.
+        default: "admin"
+    password:
+        description:
+            - Password credentials to use for auth.
+    api_key:
+        description:
+            - API key that can be used instead of I(username)/I(password) credentials.
+    operation:
```
Comment (shinmog) [inline library/panos_loopback.py]: Please remove `operation` from this module.  We're moving away from that and using `state` instead.

File: library/panos_loopback.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+    # Commit if we were asked to do so.
+    if commit:
+        try:
+            con.commit(sync=True)
+        except errors.PanDeviceError:
+            e = get_exception()
+            module.fail_json(msg='Performed {0} but commit failed: {1}'.format(operation, e.message))
+
+    # Done!
+    module.exit_json(changed=True, msg='okey dokey')
```
Comment (shinmog) [inline library/panos_loopback.py]: Heh, just say `Done` or something here :)

Comment (geraint-jones-nz) [reply library/panos_loopback.py]: `Done` :-)

File: library/panos_loopback.py (hunk 7 above)
Comment (geraint-jones-nz) [reply library/panos_loopback.py]: Okay, I have removed operation entirely from this module

File: library/panos_loopback.py (hunk 6 above)
Comment (geraint-jones-nz) [reply library/panos_loopback.py]: To my embarrassment; these were scattered throughout.. This is resolved :+1:

File: library/panos_loopback_interface.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+        raise ValueError('Virtual router {0} does not exist'.format(vr_name))
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(no_log=True),
+        username=dict(default='admin'),
+        api_key=dict(no_log=True),
+        operation=dict(default='add', choices=['add', 'update', 'delete']),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: Forgot to remove the `operation` param from argument_spec.

File: library/panos_loopback_interface.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+        state=dict(choices=['present', 'absent']),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: Give this a default of `present`

File: library/panos_loopback_interface.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+        if_name=dict(required=True),
+        ip=dict(type='list'),
+        ipv6_enabled=dict(),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: This is `type='bool'`.

File: library/panos_loopback_interface.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+        management_profile=dict(),
+        mtu=dict(),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: `type='int'`

File: library/panos_loopback_interface.py [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+        adjust_tcp_mss=dict(),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: `type='bool'`

File: library/panos_loopback_interface.py [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+        netflow_profile=dict(),
+        comment=dict(),
+        ipv4_mss_adjust=dict(),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: `type='int'`

File: library/panos_loopback_interface.py [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+        ipv6_mss_adjust=dict(),
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: `type='int'`

File: library/panos_loopback_interface.py [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+        'mtu': module.params['mtu'],
+        'adjust_tcp_mss': module.params['adjust_tcp_mss'],
+        'netflow_profile': module.params['netflow_profile'],
+        'comment': module.params['comment'],
+        'ipv4_mss_adjust': module.params['ipv4_mss_adjust'],
+        'ipv6_mss_adjust': module.params['ipv6_mss_adjust'],
+    }
+
+    # Get other info.
+    operation = module.params['operation']
```
Comment (shinmog) [inline library/panos_loopback_interface.py]: More `operation` to get rid of

File: library/panos_loopback_interface.py (hunk 9 above)
Comment (geraint-jones-nz) [reply library/panos_loopback_interface.py]: Lol, I must have been sleepy. All done

**Pull request merged**